### ISSUE TITLE
Excess property check on nested object literals dropped in 6.0 when parameter is a generic "Exact" mapped-type guard

### ISSUE THREAD

**Issue by RHOOPH** [microsoft/TypeScript#63515]
### Search Terms

excess property check, nested object literal, generic, mapped type, Exact, regression 5.9 6.0, `Options<U, T>`

### Version & Regression Information

- This changed between **5.9.3** and **6.0.x**.
- In 5.9.3 the nested excess property is reported; in 6.0 it is silently accepted.
- Reproduced on `6.0.0-dev.20260416` (see playground link) and on the released `6.0.3`.

### Playground Link

[link](https://www.typescriptlang.org/play/?ts=6.0.0-dev.20260416#code/PTAEBUGVQNgOgAygE4FMDmaDOWCWB7AOwC5QBDUdVQ1ZXAY1ACIBRADzPoBcnQBbMgAdBqACYBaLgE8RlAK5lko0Fi75BWUKjb1UOcYOTra0gFAhQ9ABap6Aa01FQAORaRwLACKh8AIwBWtlygADa4XLRkIVhwoAAU4Fa4msmgXDagAALo+PgAbqiowPjIfKAABgDyglwEhFgAPACqADSAZAQAfOWggmRcEciEcACU5sBjoJOEgmygBmkyevR0NZkArHAAnHAAzGlYjOLihPgsfOFz4qp03JN39w9ohvjiqIL0cFyacx1ayEbIRyEUC+fDpUK4GhYCZTGZzQQLEQHFZcTLwBC7faHY6nc7BI7XBj48S4dAnNAAYSIADNSSg3kZXu9PppxL9aACgaBKs4ADIATTSGQATIRRKZTNJZOxOFwGuAWqAmr8ALygADeoAA2gBpUCQ0B2VBSfDUiAAXVIeu0ETFmiNJrNTVAAH4ILrzaBSDQCshQABfADcEqlqFAFNAas1AHcXaRNWQ46BCHI+L5aAGA8HTKJbCFFGHqXJCNw6qA8rgyABxBRKZpaNi20SaCkdOJsUgy7jNRWt4akPL4XCiQOTCxUGg3UAAalAXeC6Fr4tz9HzaFARZLtScgnzkPbpAp-fLQ5HD3PF8vY7Au7IBt6yDIZTi9CIXCMIVGpgr1aXcRj8bkKQACMiqgouWCkMKmb+sMo7XqAGzbDspAciUkwAD6TOiuykM4lR-ACkyAAmEQphr4cjoKYt4GpM-6gNGgFkCBYH4BBUEwXBdwWEhuGEehoBYbAiB8Whfr3C+b4fqMQA)

### Code

```ts
type Exact<T, U> = { [K in keyof T]: K extends keyof U ? T[K] : never };

type C = { w?: { a?: number } };

declare function viaGuard<U extends C>(x: Exact<U, C>): void; // generic + Exact guard
declare function plain(x: C): void;                           // plain param (control)

viaGuard({ w: { a: 1, bogus: 2 } }); // expected: error on nested `bogus`
plain   ({ w: { a: 1, bogus: 2 } }); // error on `bogus` (both versions)
```

### Actual behavior (6.0)

Only the `plain(...)` call reports `TS2353: Object literal may only specify known properties, and 'bogus' does not exist in type '{ a?: number; }'`.

The `viaGuard(...)` call — where the argument flows through a generic type parameter `U extends C` and a mapped-type "Exact" guard `Exact<U, C>` — does **not** report the excess `bogus` key nested under `w`. The top-level `w` key is still validated; only the *nested* object literal stops being checked.

### Expected behavior

Both calls should report the excess `bogus` property, as they do in 5.9.3. Whether the parameter type is a generic Exact-guard or a plain object type should not affect excess-property checking on nested object literals.

### Additional information

- The `Exact<T, U>` mapped type is a widely-used workaround from #12936 to approximate exact object types. It still correctly maps **top-level** excess keys to `never`; the regression is specifically that **nested** object literals are no longer excess-checked under 6.0.
- The union / weak-type shape of the nested type is **not** required to reproduce — a plain `{ a?: number }`, or even `{ a: number }` with a required member, leaks identically. The minimal ingredients are: a generic function + an `Exact<U, C>` parameter + a nested object literal.
- Real-world impact: this is the exact method-signature shape used by an ORM query client — `find<U extends QueryOptions<T>>(args: Options<U, QueryOptions<T>>)` — so typos in nested `where` / `select` clauses became silently accepted after upgrading from 5.9 to 6.0.

**Comment by snarbles2** [comment]
The difference is in what `U` is inferred to be in `viaGuard<U extends C>`. It's inferred to be `C` in 5.9.3, whereas it's inferred to be ` w: { a: number, bogus: number } }` in 6.0.

I'm not sure what change resulted in the different inference, but `Exact<{w: { a: number, bogus: number } }, C>` evaluates to `w: { a: number, bogus: number } }`, so this is not a problem with EPC per se.

**Comment by RyanCavanaugh** [comment]
The reason we say exact types isn't an existing feature in TypeScript already is that the attempts to simulate it aren't exactly perfect. I don't see a bug here; two different function signatures having different behavior isn't a defect.
